database/src/postgres/state_indexer.rs

```diff
@@ -1,329 +1,438 @@
+            .execute(&self.meta_db_pool)
+            .await?;
+        Ok(())
     }
 
-    async fn delete_contract_code(
+    async fn save_state_changes_data(
         &self,
-        account_id: near_primitives::types::AccountId,
+        shard_id: near_primitives::types::ShardId,
+        state_changes: Vec<near_primitives::views::StateChangeWithCauseView>,
         block_height: u64,
         block_hash: near_indexer_primitives::CryptoHash,
     ) -> anyhow::Result<()> {
-        crate::models::StateChangesContract {
-            account_id: account_id.to_string(),
-            block_height: bigdecimal::BigDecimal::from(block_height),
-            block_hash: block_hash.to_string(),
-            data_value: None,
-        }
-        .insert_or_ignore(Self::get_connection(&self.pg_pool).await?)
-        .await
+        crate::metrics::SHARD_DATABASE_WRITE_QUERIES
+            .with_label_values(&[
+                &shard_id.to_string(),
+                "save_state_changes_data",
+                "state_changes_data",
+            ])
+            .inc();
+        let mut query_builder: sqlx::QueryBuilder<sqlx::Postgres> = sqlx::QueryBuilder::new(
+            "INSERT INTO state_changes_data (account_id, block_height, block_hash, data_key, data_value) ",
+        );
+        query_builder.push_values(state_changes.iter(), |mut values, state_change| {
+            match &state_change.value {
+                near_primitives::views::StateChangeValueView::DataUpdate {
+                    account_id,
+                    key,
+                    value,
+                } => {
+                    let data_key: &[u8] = key.as_ref();
+                    let data_value: &[u8] = value.as_ref();
+                    values
+                        .push_bind(account_id.to_string())
+                        .push_bind(bigdecimal::BigDecimal::from(block_height))
+                        .push_bind(block_hash.to_string())
+                        .push_bind(hex::encode(data_key).to_string())
+                        .push_bind(data_value);
+                }
+                near_primitives::views::StateChangeValueView::DataDeletion { account_id, key } => {
+                    let data_key: &[u8] = key.as_ref();
+                    let data_value: Option<&[u8]> = None;
+                    values
+                        .push_bind(account_id.to_string())
+                        .push_bind(bigdecimal::BigDecimal::from(block_height))
+                        .push_bind(block_hash.to_string())
+                        .push_bind(hex::encode(data_key).to_string())
+                        .push_bind(data_value);
+                }
+                _ => {}
+            }
+        });
+        query_builder.push(" ON CONFLICT DO NOTHING;");
+        query_builder
+            .build()
+            .execute(
+                self.shards_pool
+                    .get(&shard_id)
+                    .ok_or(anyhow::anyhow!("Shard not found"))?,
```

Comment:
In such error messages would be great to include the shard id we tried to simplify the debugging